src/protocol/sort/secureapplyinv.rs

```diff
@@ -43,25 +43,25 @@ impl SecureApplyInv {
     #[allow(dead_code)]
     pub async fn execute<F: Field>(
         ctx: &ProtocolContext<'_, Replicated<F>, F>,
-        input: &'_ mut Vec<Replicated<F>>,
-        sort_permutation: &'_ mut Vec<Replicated<F>>,
-    ) -> Result<(), BoxError> {
+        input: Vec<Replicated<F>>,
+        sort_permutation: Vec<Replicated<F>>,
+    ) -> Result<Vec<Replicated<F>>, BoxError> {
         let mut random_permutations =
             get_two_of_three_random_permutations(input.len(), &ctx.prss());
 
-        let (_, _) = try_join(
+        let (mut shuffled_input, sort_permutation) = try_join(
             Shuffle::new(input)
                 .execute(ctx.narrow(&ShuffleInputs), &mut random_permutations.clone()),
             Shuffle::new(sort_permutation)
                 .execute(ctx.narrow(&ShufflePermutation), &mut random_permutations),
         )
         .await?;
         let mut permutation =
-            reveal_permutation(ctx.narrow(&RevealPermutation), sort_permutation).await?;
+            reveal_permutation(ctx.narrow(&RevealPermutation), &sort_permutation).await?;
         // The paper expects us to apply an inverse on the inverted Permutation (i.e. apply_inv(permutation.inverse(), input))
         // Since this is same as apply(permutation, input), we are doing that instead to save on compute.
-        apply(&mut permutation, input);
-        Ok(())
+        apply(&mut permutation, &mut shuffled_input);
```

Comment:
is it possible for `apply` to consume `permutation`? such that this would look like
```suggestion
        apply(permutation, &mut shuffled_input);
```
?

Reply:
Took a stab at doing this